compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt

```diff
@@ -602,43 +603,61 @@ class FunctionInlining(
                     return@forEach
                 }
 
-                if (argument.isImmutableVariableLoad) {
-                    substituteMap[argument.parameter] =
-                        argument.argumentExpression.transform( // Arguments may reference the previous ones - substitute them.
-                            substitutor,
-                            data = null
-                        )
+                // Arguments may reference the previous ones - substitute them.
+                val variableInitializer = argument.argumentExpression.transform(substitutor, data = null)
+                val parameter = argument.parameter
+
+                fun substituteParameterViaTemporaryVariable() {
+                    val newVariable = createTemporaryVariable(parameter, variableInitializer, callee)
+                    evaluationStatements.add(newVariable)
+                    substituteMap[parameter] = IrGetValueWithoutLocation(newVariable.symbol)
+                }
+
+                if (alwaysCreateTemporaryVariablesForArguments && !parameter.isInlineParameter()) {
+                    substituteParameterViaTemporaryVariable()
                     return@forEach
                 }
 
-                // Arguments may reference the previous ones - substitute them.
-                val variableInitializer = argument.argumentExpression.transform(substitutor, data = null)
+                if (argument.isImmutableVariableLoad) {
+                    substituteMap[parameter] = variableInitializer
+                    return@forEach
+                }
 
                 val argumentExtracted = !argument.argumentExpression.isPure(false, context = context)
```

Comment:
For now, in fact we have to cases 
1. Directly substitute
2. Create temporary variable. 

For now logic on when which is happening is distributed all over the function code. It would be nice to collect it in one place. Like add a variable "shouldInlineArgumentValue", and just check over it and the end, without returns in the middle of function.